### ISSUE TITLE
Rule proposal: `no-unused-assign`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1693]
### Description

Got the idea from https://lgtm.com/projects/g/prettier/prettier/snapshot/dae9b0317dae184854caf66c44eec17a95200e98/files/src/language-js/parse/postprocess/typescript.js?sort=name&dir=ASC&mode=heatmap#L66

I don't think [`no-unused-vars`](https://eslint.org/docs/rules/no-unused-vars) rule should check this case, because the variable is actually "used" in some case, only should not assign.

### Fail

```js
function foo(bar) {
  bar = doSomething();
}
```

```js
function foo(bar) {
  bar = doSomething(bar);
}
```

```js
function foo(bar) {
  bar = doSomething(bar);

  // ... do something else, but not using bar
}
```

### Pass

```js
function foo(bar) {
  bar = doSomething(bar);

  return bar;
}
```

```js
function foo(bar) {
  bar = doSomething(bar);

  use(bar);
}
```

```js
function foo(bar) {
  function aInnerFunctionUsesBar() {
   return bar;
  }

  bar = doSomething(bar);

  // bar is not used but calling another function uses bar
  aInnerFunctionUsesBar()
}
```

**Comment by sindresorhus** [comment]
Accepted

**Comment by silverwind** [comment]
Ther will be false-positives in cases where the side-effect is intended, like [__webpack_public_path__](https://webpack.js.org/guides/public-path/#on-the-fly). But generally, I agree it's a good rule to have.

**Comment by fisker** [comment]
> Maybe it can detect variable type and only check func args or let bindings.

ESLint has the variable scope info to use.

**Comment by sindresorhus** [comment]
This is covered by ESLint core now. `no-useless-assignment` reports dead stores where an assigned value is never read later, and the pure write-only case is handled by `no-unused-vars`.
